### ISSUE TITLE
Built-in TypeError exception from Connection on_inbound_method()

### ISSUE THREAD

**Issue by juhamkilpinen** [celery/py-amqp#197]
There are several reports from the other projects getting this exception when using Connection class:

```
   File "/usr/lib/python2.7/site-packages/amqp/connection.py", line 472, in on_inbound_method
     return self.channels[channel_id].dispatch_method(
 TypeError: 'NoneType' object has no attribute '__getitem__'

```
See for example:
https://bugs.launchpad.net/oslo.messaging/+bug/1504725
https://github.com/celery/kombu/issues/857
https://github.com/celery/kombu/issues/463

Does this mean that Connection class is misused or should there be similar check in on_inbound_method() as there is in channel() for self.channels being None, for example:

```
def on_inbound_method(self, channel_id, method_sig, payload, content):
    if self.channels is None:
        raise RecoverableConnectionError('Connection already closed.')

```
I have seen this problem e.g. in OpenStack Pike when creating heat stacks and stack creation can fail for that TypeError exception. If I add above check it seems that client can re-connect and stack creation does not fail (when RecoverableConnectionError is thrown instead of TypeError).

**Comment by thedrow** [comment]
I'm not sure how this happens.
I'll need to see a test case that doesn't involve kombu somehow.

**Comment by juhamkilpinen** [comment]
Sorry I cannot provide a simplified test case to point the issue. I think it is somehow related to caching the connections with multiple process / threads involved, I.e. some kind of misuse of the API. I am happy for now that my workaround seems to work with the OpenStack env.

**Comment by levsh** [comment]
Hi! I got the similar error.

      File "/opt/.../embedded/lib/python3.5/site-packages/celery/app/base.py", line 745, in send_task
        amqp.send_task_message(P, name, message, **options)
      File "/opt/.../embedded/lib/python3.5/site-packages/celery/app/amqp.py", line 552, in send_task_message
        **properties
      File "/opt/.../embedded/lib/python3.5/site-packages/kombu/messaging.py", line 181, in publish
        exchange_name, declare,
      File "/opt/.../embedded/lib/python3.5/site-packages/kombu/connection.py", line 494, in _ensured
        return fun(*args, **kwargs)
      File "/opt/.../embedded/lib/python3.5/site-packages/kombu/messaging.py", line 203, in _publish
        mandatory=mandatory, immediate=immediate,
      File "/opt/.../embedded/lib/python3.5/site-packages/amqp/channel.py", line 1743, in basic_publish_confirm
        self.wait(spec.Basic.Ack)
      File "/opt/../embedded/lib/python3.5/site-packages/amqp/abstract_channel.py", line 79, in wait
        self.connection.drain_events(timeout=timeout)
      File "/opt/.../embedded/lib/python3.5/site-packages/amqp/connection.py", line 491, in drain_events
        while not self.blocking_read(timeout):
      File "/opt/.../embedded/lib/python3.5/site-packages/amqp/connection.py", line 497, in blocking_read
        return self.on_inbound_frame(frame)
      File "/opt.../embedded/lib/python3.5/site-packages/amqp/method_framing.py", line 55, in on_frame
        callback(channel, method_sig, buf, None)
      File "/opt/.../embedded/lib/python3.5/site-packages/amqp/connection.py", line 500, in on_inbound_method
        return self.channels[channel_id].dispatch_method(
    TypeError: 'NoneType' object is not subscriptable

Python 3.5.5
Python libs with some minor patches
amqp (2.3.2)
kombu (4.2.1)
celery (4.2.1)

**Comment by auvipy** [comment]
@4383 thanks!! can we chat in IRC about the usage of celery in OpenStack?

**Comment by thedrow** [comment]
Please check if #289 helps and let us know.

**Comment by 4383** [comment]
@auvipy my pleasure. 
Do not hesitate to ping me on freenode, I'm "hberaud" on freenode.

**Comment by Perepyolka** [comment]
@4383 @thedrow Hi guys, any updates on this bug? We have encountered this issue and thinking about updating our amqp to have this fix https://github.com/celery/py-amqp/pull/289.

**Comment by thedrow** [comment]
I haven't had the time to look at it.

**Comment by 4383** [comment]
@Perepyolka seems that my issue was fixed by #289
